### PR TITLE
client/core,asset: ensure coins are unlocked when no longer needed

### LINKED ISSUE #631: client/core: do not unlock outpoints on shutdown
On dexc/Core shutdown, we should definitely not unlock outpoints that were not locked by dexc/Core, and arguably we should not unlock any that are still connected to active orders.

### PR DESCRIPTION
**Pull request by itswisdomagain**
Coins are currently unlocked when
- a market order or limit order with tif immediate does not get matched
- an order is canceled (i.e. by user action)
- a market is suspended with persist=false

Other times to unlock coins (implemented in this PR):
- when new orders could not be submitted to the server, e.g. if sending the
  request or validating the server's response fails.
- when all matches are either swapped or revoked for unbooked orders, i.e.
  orders that will not receive future matches.
- after sending a swap, unlock the spent outputs to ensure accurate balance
  reporting.

~Also revoke orders in UnbookOrder note handler to ensure that coins are
unlocked for server-revoked orders. If such orders remain at status Booked,
their coins will never be unlocked because the expectation would be that the
order would receive new matches in the future.~

UPDATE: As @chappjc pointed out, revoking orders in the UnbookOrder note
handler is a hack at best, and not a very good one. The client currently has no
way of detecting revoked orders; with the implication that coins for such orders
will remain perpetually locked.

Resolves #631.

### CODE REVIEW

File: client/core/core.go [hunk 1]
```diff
@@ -2078,6 +2092,7 @@ func (c *Core) prepareTrackedTrade(dc *dexConnection, form *TradeForm, crypter e
 	}
 	err = c.db.UpdateOrder(dbOrder)
 	if err != nil {
+		// Do NOT unlock the coins, they're already locked by server.
```
Comment (itswisdomagain) [inline client/core/core.go:2095]: Scenarios like this could cause the client's subsequent orders to get rejected by the server because the client (on restart) may attempt to use any of these coins to fund another order. Since the order details could not be persisted to db, the coins would not be re-locked on restart.

One thought is to detect server errors and attempt using a different set of coins to fund the new order. I mention this because aside from this scenario, @JoeGruffins and I had experienced situations where submitting orders fail with a coin locked error from the server.

File: client/core/trade.go [hunk 2]
```diff
@@ -1541,28 +1537,73 @@ func (t *trackedTrade) processRedemption(msgID uint64, redemption *msgjson.Redem
...
-func (t *trackedTrade) hasUnswappedMatches() bool {
+// Coins will be returned if
+// - the trade status is not OrderStatusEpoch or OrderStatusBooked, that is to
+//   say, there won't be future matches for this order.
+// - there are no matches in the trade that MAY later require sending swaps,
+//   that is to say, all matches have been either swapped or revoked.
+//
+// This method modifies match fields and MUST be called with the trackedTrade
+// mutex lock held for writes.
+func (t *trackedTrade) maybeReturnCoins() bool {
```
Comment (itswisdomagain) [inline client/core/trade.go:1548]: What do we think about changing the `ReturnCoins` to `UnlockCoins` consistently. Return coins sounds as though coins were moved from the wallet but should now be returned to the wallet. Unlock sends the message that the coins were locked (to prevent unplanned spending) and should now be unlocked to permit spending anyhow.

File: client/asset/btc/btc.go [hunk 3]
```diff
@@ -970,9 +974,10 @@ func (btc *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin
 	}
 
 	// Delete the UTXOs from the cache.
-	for _, pt := range pts {
-		delete(btc.fundingCoins, pt)
+	for _, spentOp := range spentOps {
+		delete(btc.fundingCoins, spentOp.pt)
 	}
+	btc.wallet.LockUnspent(true, spentOps)
```
Comment (chappjc) [inline client/asset/btc/btc.go:980]: I believe bitoind unlocks these on spend, but I don't really object to trying it anyway.  https://github.com/bitcoin/bitcoin/pull/13160  @buck54321 Is that what you found in testing this?  @itswisdomagain if this is right and we are certain outpoints are being unlocke automatically and immediately, just replace with a comment please.

File: client/core/bookie.go [hunk 4]
```diff
@@ -408,6 +409,24 @@ func handleUnbookOrderMsg(_ *Core, dc *dexConnection, msg *msgjson.Message) erro
...
+	// to Executed or Canceled. Such orders will be revoked here, but
+	// will be marked as Executed/Canceled when the tracker receives
+	// the last set of matches in tracker.negotiate(). If the last set
+	// of matches are received before this UnbookOrder note, then the
+	// order would not be revoked because it would already have been
+	// marked as Executed or Canceled.
+	var oid order.OrderID
+	copy(oid[:], note.OrderID)
+	tracker, _, _ := dc.findOrder(oid)
+	if tracker != nil && tracker.revoke() {
```
Comment (chappjc) [inline client/core/bookie.go:426]: `unbook_order` is an `orderbook` (subscription) notification, and it should not be necessary to be subscribed to a market's `orderbook` feed for core's trade/match negotiation to function, only being auth'd with the server.

Even if that weren't the case, I'm not terribly comfortable with the ambiguities here and and switching of all unbooked orders through revoked status.  Let's try for a solution that doesn't set status as revoked for all unbooked orders.

File: client/asset/btc/btc.go (hunk 3 above)
Comment (buck54321) [reply client/asset/btc/btc.go:980]: Yes. **bitcoind** unlocks on spend. This would likely generate an error.

**Comment by chappjc** [comment]
After #513 is done and you're back to this one, can you call out any clearly-neglected coin locking sites are fixed?  There's a lot of refactoring here, so it's hard to get an idea of which specific cases are resolved.

**Comment by itswisdomagain** [comment]
Called that out in the PR description but was probably not clear. I've edited the description.
Coin-unlocking cases implemented here are:
- when new orders could not be submitted to the server, e.g. if sending the
  request or validating the server's response fails.
- when all matches are either swapped or revoked for unbooked orders, i.e.
  orders that will not receive future matches.
- after sending a swap, unlock the spent outputs to ensure accurate balance
  reporting.

Will recheck to confirm I haven't missed any, but last I checked, those were it.

File: client/asset/btc/btc.go (hunk 3 above)
Comment (itswisdomagain) [reply client/asset/btc/btc.go:980]: Simnet trading tests no longer have balance check issues, so seems both wallets are unlocking spent coins as expected. Will replace with a comment.

File: client/core/bookie.go (hunk 4 above)
Comment (itswisdomagain) [reply client/core/bookie.go:426]: I agree this is a hack, and seeing/remembering that a client is not required to subscribe to a market's orderbook even if trading in that market makes this hack insufficient at best.

I'm gonna revert/undo that change set; and we can get back to discussing how to determine revoked orders on the client so coins are unlocked.

**Comment by chappjc** [comment]
To simplify this, we can allow some cases to fall back to unlocking, possibly redundantly, on retirement of trackedTrades.  https://github.com/decred/dcrdex/pull/665

Perhaps we can simplify the approach in this PR considerably to focus on the low-hanging fruit that does not require unbook ntfn tricks.

**Comment by itswisdomagain** [comment]
> we can allow some cases to fall back to unlocking, possibly redundantly, on retirement of trackedTrades

Makes sense. To further simplify, we can make trade retirement the sole point of coin unlocking - only unlock coins when the associated trade/order is retired.

The only caveat I see there is some trade coins will remain locked for longer if a trade is kept alive simply for refund or auto-redeem reasons, as those recovery actions may take a while. We could ofc unlock coins for such orders (while not retiring them) and set `t.coinsLocked = false and t.changeLocked = false` on those trades to prevent re-attempting to unlock the coins when the trades are later retired.

If you're okay with having all coin unlocking done from the site of trade retirement, will you want to take that up in #665 or should I pull in your #665 commits and finish up here?

File: client/core/trade.go (hunk 2 above)
Comment (itswisdomagain) [reply client/core/trade.go:1712]: Err, I see. `UnlockCoins` is sort of implementation-biased. We do have a `Locked` balance in dexc (implementation-agnostic) and there are a couple other references to locking coins such as in the doc for `FundOrder`.

Probably not a big deal / head turner but sticking with Lock/Unlock wouldn't be necessarily be a reflection of a utxo-based blockchain. It'd just be harmonising expressions already used elsewhere in the codebase.

File: client/asset/btc/btc.go (hunk 3 above)
Comment (itswisdomagain) [reply client/asset/btc/btc.go:1039]: Do we know if dcr also auto-unlock coins on spend? Last, I checked, it did but I'm getting weary of trusting these asset backend behaviours when it comes to coin unlocking. We currently force coin unlocking after sending swaps for dcr, so wondering if I should as well replace that with a comment.
https://github.com/decred/dcrdex/blob/ceb666181576a60572d09bc7ff42f1c066fb598f/client/asset/dcr/dcr.go#L1027-L1031

Comment (buck54321) [reply client/asset/btc/btc.go:1039]: dcrwallet does not. We must do it manually.

**Review by buck54321 — APPROVED** [review]

**Comment by buck54321** [comment]
> If you're okay with having all coin unlocking done from the site of trade retirement

I'm pretty sure that would break things. Most importantly, it would prevent someone from using remaining funds from a canceled, but partially filled order until all matches are fully negotiated. 

Let's roll with this PR as is.

**Comment by itswisdomagain** [comment]
A breakdown of ideal coin-unlocking sites, particularly important with commit https://github.com/decred/dcrdex/pull/648/commits/48b0e22fca7a0106f24633e3b3344c2f00554925 stopping coin unlocking on client shutdown. Checked checkboxes with no link to implementation are implemented in this PR.

#### Orders for which coins should be unlocked, _copied over from https://github.com/decred/dcrdex/issues/631#issuecomment-688559845_

- [x] New orders that did not make it to the DEX, e.g. if sending the request or validating the server's response fails.
 Coins should be unlocked when an error prevents the order from reaching the DEX.
- [x] Unmatched market and tif-immediate limit orders.
 Coins are unlocked in the [`no_match` notification handler](https://github.com/decred/dcrdex/blob/ceb666181576a60572d09bc7ff42f1c066fb598f/client/core/trade.go#L281).

All other orders that receive one or more matches have no further use for locked coins when they're `Executed, Revoked or Canceled` **AND** swaps have been sent for all of the order's active (unrevoked) matches. These include partially matched market and tif-immediate limit orders (= Executed). Coin unlocking for such orders should be performed when:
  - [x] the last set of swaps are sent
  - [x] a match is revoked and the revoked match is the last match for which a swap **was** required
  - [x] the order is [canceled](url) with no swap-pending match
  - [x] the order is [revoked because the market is suspended with persist=false](https://github.com/decred/dcrdex/blob/ceb666181576a60572d09bc7ff42f1c066fb598f/client/core/trade.go#L982-L992), with no swap-pending match
  - [ ] the order is revoked by the server, with no swap-pending match

The following are cases where coins will remain needlessly locked, and resolution measures to ensure the coins are unlocked as required:
| Client is unware that... | Recover and unlock coins... |
|----|----|
| <ul><li>- [ ] a market or tif-immediate limit order was unmatched (e.g. client missed the `no_match` request) </li></ul> | |
| <ul><li>- [x] a match has been revoked (e.g. client missed the `revoke_match` request) </li></ul> | when the client (re-)connects the DEX, "notices" the revoked matches and confirms that the revoked matches are the last set of matches that would have required sending a swap |
| <ul><li>- [ ] an order has been revoked (e.g. missed the yet-to-be implemented `revoke_order` request) | |
| <ul><li>- [ ] an order has been canceled (is this possible?) | |

File: client/core/core.go (hunk 1 above)
Comment (itswisdomagain) [reply client/core/core.go:2167]: Since coins are no longer unlocked on client shutdown, the coins for this order will remain perpetually locked. The good news is the client will not attempt to re-use these coins to fund another order, which the DEX would have rejected. The bad news is that the coins will still be locked even after this order gets discarded by the DEX because the client did not respond to preimage requests. How to handle this? One thought is, since the order will not be tracked by the client, unlock the coins right away. If the client gets `coin is locked` errors from the DEX when trying to place future trades, those errors will be shortlived as the DEX will unlock those coins when the preimage request fails.

File: server/market/market.go [hunk 5]
```diff
@@ -1307,6 +1307,10 @@ func (m *Market) Unbook(lo *order.LimitOrder) bool {
 	_, removed := m.book.Remove(lo.ID())
 	m.bookMtx.Unlock()
 
+	// TODO: Should not unlock coins if there are active/unrevoked
+	// matches for this order and no swaps have been sent.
+	// If any swap has been sent, the funding coins would have been
+	// spent and it would be theoretically safe to unlock the coins.
```
Comment (chappjc) [inline server/market/market.go:1313]: I don't think this comment is right because the swapper has it's own coin locker, and the "is locked" check when processing new orders requires **both** the market's and swapper's lockers to have unlocked to consider coins unlocked (conversely, if either the swapper or market has the coin locked, it is considered locked).

Comment (chappjc) [reply server/market/market.go:1313]: Referring to https://github.com/decred/dcrdex/blob/7671c15474684ccb4e712be0de4939e66de7d3e0/server/coinlock/coinlocker.go#L53-L61

Comment (itswisdomagain) [reply server/market/market.go:1313]: Ah, gotcha. Will remove the comment.

**Comment by itswisdomagain** [comment]
I'm okay with merging this, leaving further coin unlocking work for the implementation of order status recovery described in https://github.com/decred/dcrdex/issues/597#issuecomment-689013225.

Is there any merit to temporarily restoring coin unlocking on asset shutdown? Most of those coins will be relocked on client restart, except coins for
- inactive orders (coins for these should have already been unlocked before client shutdown, or is it possible for previously active orders to be immediately considered inactive on client restart?)
- new orders submitted to the DEX but that could not be persisted to db, their coins currently remain locked until client shutdown and they're not locked on restart; could consider unlocking them right away, after all the client ditches the order right away. Only caveat to unlocking the coins right away is subsequent order placements may attempt to re-use the funding coins for the order and would get rejected by the server - until the server ditches the order (preimage request miss) and unlocks the coin, that is.
https://github.com/decred/dcrdex/blob/7671c15474684ccb4e712be0de4939e66de7d3e0/client/core/core.go#L2151-L2155

In summary, I don't think client shutdown should be the trigger for unlocking coins. Only unlock coins when they're no longer needed, falling back to unlocking coins when trades are retired. As long as the client considers a trade active, the associated coins should remain locked.

I'm probably gonna want to move the client coin unlocking requirement described in https://github.com/decred/dcrdex/pull/648#issuecomment-688585289 to an issue to ensure coins don't stay locked against expectations.

File: client/core/core.go [hunk 6]
```diff
@@ -2117,6 +2128,8 @@ func (c *Core) prepareTrackedTrade(dc *dexConnection, form *TradeForm, crypter e
 	result := new(msgjson.OrderResult)
 	err = dc.signAndRequest(msgOrder, route, result, DefaultResponseTimeout)
 	if err != nil {
+		// Do NOT unlock the coins because the request may have actually reached
+		// the server.
 		return nil, 0, fmt.Errorf("new order request with DEX server %v failed: %w", dc.acct.host, err)
```
Comment (itswisdomagain) [inline client/core/core.go:2133]: I'm reconsidering this. The client does not save the order to db, nor will it respond to preimage requests for the order. Only advantage to keeping the coins locked is to prevent new orders from using these coins as they're already locked by the server. But we can't keep these locked forever. And there's no client-restart-agnostic reliable way of knowing that the server has also ditched the order, so we can safely unlock coins.

Any objections to unlocking right away?

File: client/core/core.go (hunk 1 above)
Comment (itswisdomagain) [reply client/core/core.go:2167]: Previous comment applies here as well:

The client does not save the order to db, nor will it respond to preimage requests for the order. Only advantage to keeping the coins locked is to prevent new orders from using these coins as they're already locked by the server. But we can't keep these locked forever. And there's no client-restart-agnostic reliable way of knowing that the server has also ditched the order, so we can safely unlock coins.

Any objections to unlocking right away?

File: client/core/core.go [hunk 7]
```diff
@@ -3234,17 +3256,10 @@ out:
...
-					log.Infof("Retiring inactive order %v", oid)
-					trade.mtx.RLock()
-					if trade.change != nil {
-						changeCoins := asset.Coins{trade.change}
-						err := trade.wallets.fromWallet.ReturnCoins(changeCoins)
-						log.Debugf("Unlocked order %v change coins: %v (wallet error = %v)", oid, changeCoins, err)
-					}
-					tradeCoins := trade.coinList()
-					err := trade.wallets.fromWallet.ReturnCoins(tradeCoins)
-					log.Debugf("Unlocked funding coins for order %v: %v (wallet error = %v)", oid, tradeCoins, err)
```
Comment (itswisdomagain) [inline client/core/core.go:3246]: @chappjc were these intentional for debugging purposes? It's likely/expected that the funding coins or change coin for inactive orders have been returned already. Also, there'd be little need to unlock funding coins if there's a change coin (the funding coin must have been spent to create the change coin, and thus unlocked).

I've replaced these with `trade.returnCoins()` which performs selective coin unlocking. If there are issues with coin unlocking, your approach of unlocking everything and logging the response will reveal the issues. Otherwise, it'd just be spammy. What's your take?

File: client/core/trade.go [hunk 8]
```diff
@@ -1688,28 +1700,75 @@ func (t *trackedTrade) processRedemption(msgID uint64, redemption *msgjson.Redem
...
-		coins = append(coins, coin)
+// returnCoins unlocks this trade's funding coins (if unspent) or the change
+// coin if a previous swap created a change coin that is locked.
+// Coins are auto-unlocked once spent in a swap tx, including intermediate
+// change coins, such that only the last change coin (if locked), will need
+// to be unlocked.
+//
+// This method modifies match fields and MUST be called with the trackedTrade
+// mutex lock held for writes.
+func (t *trackedTrade) returnCoins() {
```
Comment (itswisdomagain) [inline client/core/trade.go:1752]: Will need a way to communicate to `Core` to refresh the fromWallet balance after unlocking coins. Probably a better candidate for #672. Will add a comment there.

File: client/core/core.go (hunk 7 above)
Comment (chappjc) [reply client/core/core.go:3246]: That's fine.  This was just the least clever and most thorough solution.  Never intended as the best solution, just a band-aid.

**Review by chappjc — APPROVED** [review]

File: client/core/core.go (hunk 6 above)
Comment (chappjc) [reply client/core/core.go:2133]: This is an odd case.  It's hard to imagine the request failing but it actually being booked, but if the client cannot do anything with the order, I don't see a reason to keep the coins locked.

File: client/core/trade.go (hunk 8 above)
Comment (chappjc) [reply client/core/trade.go:1752]: The caller of `returnCoins` be doing this, as in `tick`?  We're relying on tick to do this now though it seems

The DB failure with `UpdateOrder` is hard to handle because it basically means the program is totally hosed.  I don't have a good resolution in mind.

Comment (itswisdomagain) [reply client/core/trade.go:1752]: `returnCoins` may be called from a number of places aside `tick`. I'm not sure how I feel about having all callers of `returnCoins` also refresh wallet balance but will think this through more when working on #672.
